src/indexer/index_writer_status.rs

```diff
@@ -0,0 +1,118 @@
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, RwLock};
+
+use super::AddBatchReceiver;
+
+#[derive(Clone)]
+pub(crate) struct IndexWriterStatus {
+    inner: Arc<Inner>,
+}
+
+impl IndexWriterStatus {
+    /// Returns true iff the index writer is alive.
+    pub fn is_alive(&self) -> bool {
+        self.inner.as_ref().is_alive()
+    }
+
+    /// Returns a copy of the operation receiver.
+    /// If the index writer was killed, returns None.
+    pub fn operation_receiver(&self) -> Option<AddBatchReceiver> {
+        let rlock = self
+            .inner
+            .receive_channel
+            .read()
+            .expect("This lock should never be poisoned");
+        rlock.as_ref().map(|receiver| receiver.clone())
+    }
+
+    /// Create an index writer bomb.
+    /// If dropped, the index writer status will be killed.
+    pub(crate) fn create_bomb(&self) -> IndexWriterBomb {
+        IndexWriterBomb {
+            inner: Some(self.inner.clone()),
+        }
+    }
+}
+
+struct Inner {
+    is_alive: AtomicBool,
+    receive_channel: RwLock<Option<AddBatchReceiver>>,
+}
+
+impl Inner {
+    fn is_alive(&self) -> bool {
+        self.is_alive.load(Ordering::Relaxed)
+    }
+
+    fn kill(&self) {
+        self.is_alive.store(false, Ordering::Relaxed);
+        self.receive_channel
+            .write()
+            .expect("This lock should never be poisoned")
+            .take();
+    }
+}
+
+impl From<AddBatchReceiver> for IndexWriterStatus {
+    fn from(receiver: AddBatchReceiver) -> Self {
+        IndexWriterStatus {
+            inner: Arc::new(Inner {
+                is_alive: AtomicBool::new(true),
+                receive_channel: RwLock::new(Some(receiver)),
+            }),
+        }
+    }
+}
+
+/// If dropped, the index writer will be killed.
+/// To prevent this, clients can call `.defuse()`.
+pub(crate) struct IndexWriterBomb {
+    inner: Option<Arc<Inner>>,
+}
+
+impl IndexWriterBomb {
+    /// Defuses the bomb.
+    ///
+    /// This is the only way to drop the bomb without killing
+    /// the index writer.
+    pub fn defuse(mut self) {
```

Comment:
I really like this kind of analogy, it makes the code immediately understandable.